File: 2025/day08/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Point struct {
	x, y, z int
}

type UnionFind struct {
	parent []int
	rank   []int
	size   []int
}

func NewUnionFind(n int) *UnionFind {
	uf := &UnionFind{
		parent: make([]int, n),
		rank:   make([]int, n),
		size:   make([]int, n),
	}
	for i := 0; i < n; i++ {
		uf.parent[i] = i
		uf.size[i] = 1
	}
	return uf
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // Path compression
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) bool {
	px, py := uf.Find(x), uf.Find(y)
	if px == py {
		return false // Already in same set
	}

	// Union by rank
	if uf.rank[px] < uf.rank[py] {
		px, py = py, px
	}
	uf.parent[py] = px
	uf.size[px] += uf.size[py]
	if uf.rank[px] == uf.rank[py] {
		uf.rank[px]++
	}
	return true
}

func (uf *UnionFind) GetComponentSizes() []int {
	sizes := []int{}
	for i := 0; i < len(uf.parent); i++ {
		if uf.parent[i] == i { // Root of a component
			sizes = append(sizes, uf.size[i])
		}
	}
	return sizes
}

type Pair struct {
	distSq int
	i, j   int
}

func euclideanDistanceSq(p1, p2 Point) int {
	dx := p1.x - p2.x
	dy := p1.y - p2.y
	dz := p1.z - p2.z
	return dx*dx + dy*dy + dz*dz
}

func parseInput(filename string) ([]Point, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	points := []Point{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) != 3 {
			continue
		}
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		z, _ := strconv.Atoi(parts[2])
		points = append(points, Point{x, y, z})
	}
	return points, scanner.Err()
}

func part1(points []Point, numConnections int) int {
	n := len(points)

	// Generate all pairs with distances
	pairs := []Pair{}
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			distSq := euclideanDistanceSq(points[i], points[j])
			pairs = append(pairs, Pair{distSq, i, j})
		}
	}

	// Sort by distance
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].distSq < pairs[j].distSq
	})

	// Union-Find to connect closest pairs
	uf := NewUnionFind(n)
	connections := 0
	for _, pair := range pairs {
		uf.Union(pair.i, pair.j)
		connections++
		if connections == numConnections {
			break
		}
	}

	// Get component sizes and find the 3 largest
	sizes := uf.GetComponentSizes()
	sort.Slice(sizes, func(i, j int) bool {
		return sizes[i] > sizes[j]
	})

	// Multiply the 3 largest
	return sizes[0] * sizes[1] * sizes[2]
}

func part2(points []Point) int {
	n := len(points)

	// Generate all pairs with distances
	pairs := []Pair{}
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			distSq := euclideanDistanceSq(points[i], points[j])
			pairs = append(pairs, Pair{distSq, i, j})
		}
	}

	// Sort by distance
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].distSq < pairs[j].distSq
	})

	// Union-Find to connect until all in one circuit
	uf := NewUnionFind(n)
	numComponents := n

	for _, pair := range pairs {
		if uf.Union(pair.i, pair.j) { // Actually merged two components
			numComponents--
			if numComponents == 1 {
				// This was the last connection - all in one circuit now
				return points[pair.i].x * points[pair.j].x // Product of X coordinates
			}
		}
	}

	return 0
}

func main() {
	inputFile := "../input.txt"
	if len(os.Args) > 1 {
		inputFile = os.Args[1]
	}

	points, err := parseInput(inputFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(points, 1000))
	fmt.Printf("Part 2: %d\n", part2(points))
}

File: 2025/day08/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_POINTS 10000

typedef struct {
    int x, y, z;
} Point;

typedef struct {
    long long dist_sq;
    int i, j;
} Pair;

typedef struct {
    int *parent;
    int *rank;
    int *size;
    int n;
} UnionFind;

// Union-Find functions
UnionFind* uf_create(int n) {
    UnionFind *uf = malloc(sizeof(UnionFind));
    uf->n = n;
    uf->parent = malloc(n * sizeof(int));
    uf->rank = malloc(n * sizeof(int));
    uf->size = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++) {
        uf->parent[i] = i;
        uf->rank[i] = 0;
        uf->size[i] = 1;
    }

    return uf;
}

void uf_free(UnionFind *uf) {
    free(uf->parent);
    free(uf->rank);
    free(uf->size);
    free(uf);
}

int uf_find(UnionFind *uf, int x) {
    if (uf->parent[x] != x) {
        uf->parent[x] = uf_find(uf, uf->parent[x]);  // Path compression
    }
    return uf->parent[x];
}

int uf_union(UnionFind *uf, int x, int y) {
    int px = uf_find(uf, x);
    int py = uf_find(uf, y);

    if (px == py) {
        return 0;  // Already in same set
    }

    // Union by rank
    if (uf->rank[px] < uf->rank[py]) {
        int temp = px;
        px = py;
        py = temp;
    }

    uf->parent[py] = px;
    uf->size[px] += uf->size[py];

    if (uf->rank[px] == uf->rank[py]) {
        uf->rank[px]++;
    }

    return 1;  // Successfully merged
}

void uf_get_component_sizes(UnionFind *uf, int *sizes, int *count) {
    *count = 0;
    for (int i = 0; i < uf->n; i++) {
        if (uf->parent[i] == i) {  // Root of a component
            sizes[(*count)++] = uf->size[i];
        }
    }
}

// Calculate squared Euclidean distance
long long dist_sq(Point *p1, Point *p2) {
    long long dx = (long long)(p1->x - p2->x);
    long long dy = (long long)(p1->y - p2->y);
    long long dz = (long long)(p1->z - p2->z);
    return dx * dx + dy * dy + dz * dz;
}

// Comparison function for qsort
int compare_pairs(const void *a, const void *b) {
    Pair *pa = (Pair *)a;
    Pair *pb = (Pair *)b;
    if (pa->dist_sq < pb->dist_sq) return -1;
    if (pa->dist_sq > pb->dist_sq) return 1;
    return 0;
}

// Comparison function for sorting sizes in descending order
int compare_desc(const void *a, const void *b) {
    return (*(int *)b - *(int *)a);
}

// Parse input file
int parse_input(const char *filename, Point *points) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("Error opening file");
        return -1;
    }

    int count = 0;
    char line[256];
    while (fgets(line, sizeof(line), f)) {
        if (sscanf(line, "%d,%d,%d", &points[count].x, &points[count].y, &points[count].z) == 3) {
            count++;
        }
    }

    fclose(f);
    return count;
}

long long part1(Point *points, int n, int num_connections) {
    // Generate all pairs with distances
    int num_pairs = n * (n - 1) / 2;
    Pair *pairs = malloc(num_pairs * sizeof(Pair));

    int idx = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            pairs[idx].dist_sq = dist_sq(&points[i], &points[j]);
            pairs[idx].i = i;
            pairs[idx].j = j;
            idx++;
        }
    }

    // Sort by distance
    qsort(pairs, num_pairs, sizeof(Pair), compare_pairs);

    // Union-Find to connect closest pairs
    UnionFind *uf = uf_create(n);
    int connections = 0;

    for (int k = 0; k < num_pairs && connections < num_connections; k++) {
        uf_union(uf, pairs[k].i, pairs[k].j);
        connections++;
    }

    // Get component sizes and find the 3 largest
    int *sizes = malloc(n * sizeof(int));
    int num_components;
    uf_get_component_sizes(uf, sizes, &num_components);

    // Sort sizes in descending order
    qsort(sizes, num_components, sizeof(int), compare_desc);

    // Multiply the 3 largest
    long long result = (long long)sizes[0] * sizes[1] * sizes[2];

    free(sizes);
    free(pairs);
    uf_free(uf);

    return result;
}

long long part2(Point *points, int n) {
    // Generate all pairs with distances
    int num_pairs = n * (n - 1) / 2;
    Pair *pairs = malloc(num_pairs * sizeof(Pair));

    int idx = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            pairs[idx].dist_sq = dist_sq(&points[i], &points[j]);
            pairs[idx].i = i;
            pairs[idx].j = j;
            idx++;
        }
    }

    // Sort by distance
    qsort(pairs, num_pairs, sizeof(Pair), compare_pairs);

    // Union-Find to connect until all in one circuit
    UnionFind *uf = uf_create(n);
    int num_components = n;
    long long result = 0;

    for (int k = 0; k < num_pairs; k++) {
        if (uf_union(uf, pairs[k].i, pairs[k].j)) {
            num_components--;
            if (num_components == 1) {
                // This was the last connection
                result = (long long)points[pairs[k].i].x * points[pairs[k].j].x;
                break;
            }
        }
    }

    free(pairs);
    uf_free(uf);

    return result;
}

int main(int argc, char *argv[]) {
    const char *filename = argc > 1 ? argv[1] : "../input.txt";

    Point *points = malloc(MAX_POINTS * sizeof(Point));
    int n = parse_input(filename, points);

    if (n < 0) {
        fprintf(stderr, "Failed to parse input\n");
        free(points);
        return 1;
    }

    printf("Part 1: %lld\n", part1(points, n, 1000));
    printf("Part 2: %lld\n", part2(points, n));

    free(points);
    return 0;
}
